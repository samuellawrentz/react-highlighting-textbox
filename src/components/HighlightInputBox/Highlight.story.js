import React from 'react';
import HighlightInputBox from './index';

export default {
  title: 'Hightlight Inputbox',
  component: HighlightInputBox,
};

const Template = (args) => <HighlightInputBox {...args} />;

export const Basic = Template.bind({});
Basic.args = {
  word: 'test',
  triggerOn: 'change',
  value: 'This is a test',
  onChange: (value) => console.log(value),
};
