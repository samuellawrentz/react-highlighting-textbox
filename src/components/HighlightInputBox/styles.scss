.sam-react-highlight-span {
  color: red;
}
.sam-react-highlight-inputbox[contenteditable='true'] {
  white-space: nowrap;
  background: white;
  overflow: hidden;
  outline: none;
  width: 100%;
}
.sam-react-highlight-inputbox[contenteditable='true'] br {
  display: none;
}
.sam-react-highlight-inputbox[contenteditable='true'] span {
  display: inline;
  white-space: nowrap;
}
