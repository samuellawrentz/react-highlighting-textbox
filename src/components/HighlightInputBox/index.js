import React from 'react';
import ContentEditable from 'react-contenteditable';
import './styles.scss';

/**
 * A React component that renders a ContentEditable div with the ability to highlight specific words or phrases.
 */
class HighlightInputBox extends React.Component {
  /**
   * Initializes the state of the component and creates a ref for the ContentEditable div.
   * @param {Object} props - The component props.
   * @param {string} props.value - The value of the input box.
   * @param {string | string[]} props.word - The word(s) to highlight.
   * @param {string} props.highlightClassName - The class name to apply to the span element that highlights the word(s).
   * @param {Function} props.onChange - The callback function to trigger when the value of the input box changes.
   * @param {Function} props.onBlur - The callback function to trigger when the user clicks outside of the input box.
   * @param {Function} props.onKeyUp - The callback function to trigger when the user releases a key.
   * @param {Function} props.onClick - The callback function to trigger when the user clicks on the input box.
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = { html: this.getHTML(props.value) };
    this.ref = React.createRef();
  }

  /**
   * Replaces a specified word in a string with an HTML span element that has a specified class and an onClick event that selects the span element.
   * @param {string} word - The word to highlight.
   * @param {string} textContent - The text content to search for the word.
   * @returns {string} - The modified text content with the word highlighted.
   */
  replacer = (word, textContent) => {
    // Don't highlight the word if it is too long
    if (word.length > 40) return textContent;
    const rg = new RegExp(`${word}`, 'gi');
    const spanClassName = this.props.highlightClassName || 'sam-react-highlight-span';

    return textContent.replace(
      rg,
      `<span onClick="var r=document.createRange();r.selectNode(this);window.getSelection().removeAllRanges();window.getSelection().addRange(r)" class="${spanClassName}">${word}</span>`,
    );
  };

  /**
   * Returns the HTML content with the specified word(s) highlighted.
   * @param {string} value - The value of the input box.
   * @returns {string} - The HTML content with the specified word(s) highlighted.
   */
  getHTML = (value) => {
    if (!value) return value;
    const { word } = this.props;
    const textContent = value || (this.ref && this.ref.current ? this.ref.current.textContent : '');
    if (!word) return textContent;
    let htmlContent = textContent;
    if (typeof word === 'string') htmlContent = this.replacer(word, htmlContent);
    else if (Array.isArray(word))
      word.forEach((w) => {
        htmlContent = this.replacer(w, htmlContent);
      });

    return htmlContent;
  };

  /**
   * Updates the state of the component with the highlighted HTML content and triggers a callback function when the user clicks outside of the div.
   * @param {Event} evt - The event object.
   * @param {string} value - The value of the input box.
   * @returns {void}
   */
  handleChange = (evt) => {
    evt.persist?.();
    const value = this.props.triggerOn === 'change' ? evt.target.value : this.ref.current.textContent;
    this.setState({ html: this.getHTML(value) }, () => {
      const e = Object.assign({}, evt, { target: { value } });
      if (this.props.onChange && this.ref && this.ref.current) this.props.onChange(e);
    });
  };

  /**
   * Updates the state of the component when the props change.
   * @param {Object} prevProps - The previous props.
   * @param {Object} prevState - The previous state.
   * @returns {void}
   */
  componentDidUpdate(prevProps) {
    if (prevProps.value !== this.props.value || prevProps.word !== this.props.word) {
      this.setState({ html: this.getHTML(this.props.value) });
    }
  }

  /**
   * Renders the HighlightInputBox component.
   * @returns {JSX.Element} - The rendered component.
   */
  render() {
    return (
      <ContentEditable
        className="sam-react-highlight-inputbox"
        innerRef={this.ref}
        html={this.state.html} // innerHTML of the editable div
        disabled={false} // use true to disable edition
        onChange={this.props.triggerOn === 'change' ? this.handleChange : undefined} // handle innerHTML change
        onBlur={this.props.triggerOn === 'blur' ? this.handleChange : undefined} // handle innerHTML change
        onKeyUp={this.props.onKeyUp}
        onClick={this.props.onClick}
      />
    );
  }
}

export default React.memo(React.forwardRef((props, ref) => <HighlightInputBox innerRef={ref} {...props} />));
