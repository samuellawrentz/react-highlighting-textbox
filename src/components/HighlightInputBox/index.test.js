// Generated by CodiumAI
import React from 'react';
import HighlightInputBox from './index';
import { shallow, mount } from 'enzyme';
// Generated by CodiumAI

describe('HighlightInputBox_class', () => {
  // Tests that the HighlightInputBox component renders with default props
  it('test_default_props', () => {
    const wrapper = shallow(<HighlightInputBox />);
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the HighlightInputBox component renders with custom props
  it('test_custom_props', () => {
    const wrapper = shallow(
      <HighlightInputBox
        value="Hello World"
        word="Hello"
        highlightClassName="highlight"
        onChange={() => {}}
        onBlur={() => {}}
        onKeyUp={() => {}}
        onClick={() => {}}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the word is not highlighted if it is too long
  it('test_word_too_long', () => {
    const wrapper = shallow(<HighlightInputBox value="Hello World" word="ThisIsAVeryLongWordThatShouldNotBeHighlighted" />);
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the word is not highlighted if it is not found in the input box
  it('test_word_not_found', () => {
    const wrapper = shallow(<HighlightInputBox value="Hello World" word="Goodbye" />);
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the word is not highlighted if it is an empty string
  it('test_empty_word', () => {
    const wrapper = shallow(<HighlightInputBox value="Hello World" word="" />);
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the HTML content is not updated if the value of the input box is null
  it('test_null_value', () => {
    const wrapper = shallow(<HighlightInputBox value={null} word="Hello" />);
    expect(wrapper).toMatchSnapshot();
  });

  // Tests that the word is highlighted when it is a number
  it('test_word_highlight_number', () => {
    const wrapper = shallow(<HighlightInputBox value="12345" word="12345" />);
    expect(wrapper.find('.sam-react-highlight-span')).toHaveLength(1);
  });

  // Tests that the HighlightInputBox component correctly highlights an array of numbers
  it('test_word_highlight_array', () => {
    const wrapper = mount(<HighlightInputBox value="" word={[1, 2, 3]} />);
    const htmlContent = wrapper.find('ContentEditable').prop('html');
    expect(htmlContent).toContain('<span class="sam-react-highlight-span">1</span>');
    expect(htmlContent).toContain('<span class="sam-react-highlight-span">2</span>');
    expect(htmlContent).toContain('<span class="sam-react-highlight-span">3</span>');
  });

  // Tests that the HTML content is empty when the value of the input box is undefined
  it('test_value_undefined', () => {
    const wrapper = mount(<HighlightInputBox value={undefined} />);
    expect(wrapper.find('ContentEditable').prop('html')).toBe('');
  });
});
