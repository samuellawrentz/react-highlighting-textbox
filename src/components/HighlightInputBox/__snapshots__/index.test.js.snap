// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightInputBox_class test_custom_props 1`] = `ShallowWrapper {}`;

exports[`HighlightInputBox_class test_default_props 1`] = `ShallowWrapper {}`;

exports[`HighlightInputBox_class test_empty_word 1`] = `ShallowWrapper {}`;

exports[`HighlightInputBox_class test_null_value 1`] = `ShallowWrapper {}`;

exports[`HighlightInputBox_class test_word_not_found 1`] = `ShallowWrapper {}`;

exports[`HighlightInputBox_class test_word_too_long 1`] = `ShallowWrapper {}`;
